### PR TITLE
Improve gitignore parsing

### LINKED ISSUE #202: Gitignore is not respected for files outside working directory that are not in a git repository
Running OSV-Scanner on a file/directory that is not in a git repository (or is not a git repository itself) currently parses gitignore files starting from the current working directory where OSV-Scanner was invoked.

This means that if a OSV-Scanner is run on a non-git file that is not under the current working directory (e.g. `osv-scanner -r ../foo/`), any gitignore files under the target will not be parsed.

This only affects non-git files/directories. If the target file/directory is part of a git repository then OSV-Scanner will parse gitignore files starting from the root of the file's containing git repository, as intended.

### PR DESCRIPTION
**Pull request by michaelkedar**
I had thought of some minor improvements for the gitignore parsing I had implemented:
- Parse gitignores for non-git files from their respective directories, instead of the current working directory.
  - Should fix #202, and I think also a potential issue on Windows scanning across drive letters.
- Fix some error messages not having newlines.
- Add a error message warning if the root file is ignored, instead of skipping silently.
- Add some notes in USAGE.md about gitignore behaviour.

### CODE REVIEW

File: pkg/osvscanner/osvscanner.go [hunk 1]
```diff
@@ -124,9 +128,23 @@ type gitIgnoreMatcher struct {
 
 func parseGitIgnores(dir string) (*gitIgnoreMatcher, error) {
 	// We need to parse .gitignore files from the root of the git repo to correctly identify ignored files
-	// Defaults to current directory if dir is not in a repo or some other error
-	// TODO: Won't parse ignores if dir is not in a git repo, and is not under the current directory (e.g ../path/to)
-	fs := osfs.New(".")
+	var fs billy.Filesystem
+
+	// Default to dir (or directory containing dir if it's a file) is not in a repo or some other error
+	f, err := os.Open(dir)
```
Comment (another-rex) [inline pkg/osvscanner/osvscanner.go:134]: I think you can combine the Open and Stat into `os.Stat`

**Review by another-rex — APPROVED** [review]
Thanks!, Looks good just one comment. Also the USAGE.md changes will clash with the documentation moving that is being done in #201, so if that get's merged in first the docs will have to be moved.

File: pkg/osvscanner/osvscanner.go (hunk 1 above)
Comment (michaelkedar) [reply pkg/osvscanner/osvscanner.go]: Oh, I didn't see that function. Changed.

**Comment by michaelkedar** [comment]
Re: #201 / #203
I don't mind waiting until that's merged and fixing it here. Or we can merge this now and fix the usage there.
Up to you.

**Comment by another-rex** [comment]
That docs PR might take a bit longer, let's just merge this now.

File: USAGE.md [hunk 2]
```diff
@@ -30,6 +30,13 @@ The recursive flag `-r` or `--recursive` will tell the scanner to search all sub
 
 Git directories are searched for the latest commit hash. Searching for git commit hash is intended to work with projects that use git submodules or a similar mechanism where dependencies are checked out as real git repositories. 
 
+### Ignored files
+
+By default, OSV-Scanner will not scan files that are ignored by `.gitignore` files. If the specified file is not part of a git repository, `.gitignore` files are parsed recursively starting from the target directory,
```
Comment (oliverchang) [inline USAGE.md:35]: "recursively" seems a bit vague -- can you clarify what this means? 

That said -- does the behaviour actually make sense? If a file isn't part of a git repo, no gitignores should apply to it right?

File: USAGE.md [hunk 3]
```diff
@@ -30,6 +30,13 @@ The recursive flag `-r` or `--recursive` will tell the scanner to search all sub
...
+otherwise they are parsed from the root of the git repository as typical.
+
+The `--no-ignore` flag can be used force the scanner to scan ignored files.
```
Comment (oliverchang) [inline USAGE.md:38]: "can be used **to** force the scannner"

File: pkg/osvscanner/osvscanner.go [hunk 4]
```diff
@@ -124,9 +128,19 @@ type gitIgnoreMatcher struct {
 
 func parseGitIgnores(dir string) (*gitIgnoreMatcher, error) {
 	// We need to parse .gitignore files from the root of the git repo to correctly identify ignored files
-	// Defaults to current directory if dir is not in a repo or some other error
-	// TODO: Won't parse ignores if dir is not in a git repo, and is not under the current directory (e.g ../path/to)
-	fs := osfs.New(".")
+	var fs billy.Filesystem
+
+	// Default to dir (or directory containing dir if it's a file) is not in a repo or some other error
+	finfo, err := os.Stat(dir)
```
Comment (oliverchang) [inline pkg/osvscanner/osvscanner.go:134]: Can we rename `dir` to `path` to avoid the confusing variable name?

File: USAGE.md (hunk 2 above)
Comment (oliverchang) [reply USAGE.md:35]: CC @another-rex

Comment (michaelkedar) [reply USAGE.md:35]: "recursively" meaning it parses `foo/.gitignore`, `foo/bar/.gitignore`, `foo/bar/baz/.gitignore`, etc.
I guess it'd be more clear to say "`.gitignore` files are parsed in the target directory and all subdirectories"?

I think this behaviour is the probably the most useful:
- I don't think you'd typically have `.gitignore` files in directories that aren't git repositories.
   - If you do, I would assume that you'd want to use it.
- Say you have a folder `foo/`, that's not a repo, and inside have `foo/bar/` and `foo/baz/` that are git repos. Running `osv-scanner -r foo/` should pick up the gitignore files from `bar/` and `baz/`.

File: USAGE.md (hunk 3 above)
Comment (michaelkedar) [reply USAGE.md:38]: :upside_down_face: I guess I'll make another PR to fix this

File: USAGE.md (hunk 2 above)
Comment (oliverchang) [reply USAGE.md:35]: The way this is worded was rather unclear. It also doesn't clearly distinguish "specified file" from "target directory", and it wasn't clear what "not part of a git repository" means exactly. 

Would this alternative wording work:  

"All recursively scanned files are matched to a git repository (if it exists) and any corresponding .gitignore files are taken into account."

So just to make sure my own understanding is correct, using your example:  

Given `foo/` (not a git repo), `foo/bar` (git repo), and `foo/baz` (git repo):
- Scanning `foo/bar/blah` would use `foo/bar/.gitignore` if it exists
- Scanning `foo/baz/blah` would use `foo/baz/.gitignore` if it exists
- Scanning `foo/blah` would **also** use `foo/.gitignore` if it exists, even though it's not a git repo? 

There are variations of edge cases here. e.g. 
- If `foo/` is not a git repo AND has a gitignore,, and `foo/bar` is a git repo does not have a git ignore, does scannign something in `foo/bar` use the gitignore from `foo/` ?

Comment (michaelkedar) [reply USAGE.md:35]: > So just to make sure my own understanding is correct, using your example:

These are all correct, as it currently works.

> Would this alternative wording work:
>
>"All recursively scanned files are matched to a git repository (if it exists) and any matching .gitignore files within that repository are taken into account."

I suppose, though I don't think it captures the (current) non-repository behaviour (e.g. dot point 3)

Comment (oliverchang) [reply USAGE.md:35]: > > So just to make sure my own understanding is correct, using your example:
> 
> These are all correct, as it currently works.
> 
> > Would this alternative wording work:
> > "All recursively scanned files are matched to a git repository (if it exists) and any matching .gitignore files within that repository are taken into account."
> 
> I suppose, though I don't think it captures the (current) non-repository behaviour (e.g. dot point 3)

There's a bunch of edge cases (e.g. the last one I listed in my last comment) that would be resolved if we more closely followed git behaviour (which should take into account repo boundaries). Is that easy to support?

Comment (michaelkedar) [reply USAGE.md:35]: Tracking repo boundaries correctly is a bit tricky. I've made the bug #209 for this.

File: USAGE.md (hunk 3 above)
Comment (michaelkedar) [reply USAGE.md:38]: Fixed in #208

File: pkg/osvscanner/osvscanner.go (hunk 4 above)
Comment (michaelkedar) [reply pkg/osvscanner/osvscanner.go:134]: Fixed in #208

**Pull request merged**